core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -450,23 +459,27 @@ class TransactionCoordinator(brokerId: Int,
             Left(Errors.INVALID_PRODUCER_ID_MAPPING)
           } else {
             val txnMetadata = epochAndTxnMetadata.transactionMetadata
-            txnMetadata.inLock(txnMetadata.fenceProducerEpoch())
-            handleEndTransaction(txnMetadata.transactionalId,
-              txnMetadata.producerId,
-              txnMetadata.producerEpoch,
-              TransactionResult.ABORT,
-              {
-                case Errors.NONE =>
-                  info(s"Completed rollback ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} due to timeout")
-                case e @ (Errors.INVALID_PRODUCER_ID_MAPPING |
-                          Errors.INVALID_PRODUCER_EPOCH |
-                          Errors.CONCURRENT_TRANSACTIONS) =>
-                  debug(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} has aborted due to ${e.exceptionName}")
-                case e =>
-                  warn(s"Rolling back ongoing transaction of transactionalId: ${txnIdAndPidEpoch.transactionalId} failed due to ${e.exceptionName}")
-              })
-            Right(epochAndTxnMetadata)
-          }
+            if (txnMetadata.pendingTransitionInProgress) {
```

Comment:
I think we should have the inLock covering the first condition as well since it is reading the pending state right?